rand_distr/src/hypergeometric.rs

```diff
@@ -0,0 +1,404 @@
+                        let yn = n1 as f64 - y + 1.0;
+                        let yk = k as f64 - y + 1.0;
+                        let nk = n2 as f64 - k as f64 + y1;
+                        let r = -ym / y1;
+                        let s = ym / yn;
+                        let t = ym / yk;
+                        let e = -ym / nk;
+                        let g = yn * yk / (y1 * nk) - 1.0;
+                        let dg = if g < 0.0 {
+                            1.0 + g
+                        } else {
+                            g
+                        };
+                        let gu = g * (1.0 + g * (-0.5 + g / 3.0));
+                        let gl = gu - g.powi(4) / (4.0 * dg);
+                        let xm = m + 0.5;
+                        let xn = n1 as f64 - m + 0.5;
+                        let xk = k as f64 - m + 0.5;
+                        let nm = n2 as f64 - k as f64 + xm;
+                        let ub = xm * r * (1.0 + r * (-0.5 + r / 3.0)) +
+                            xn * s * (1.0 + s * (-0.5 + s / 3.0)) +
+                            xk * t * (1.0 + t * (-0.5 + t / 3.0)) +
+                            nm * e * (1.0 + e * (-0.5 + e / 3.0)) +
+                            y * gu - m * gl + 0.0034;
+                        let av = v.ln();
+                        if av > ub { continue; }
+                        let dr = if r < 0.0 {
+                            xm * r.powi(4) / (1.0 + r)
+                        } else {
+                            xm * r.powi(4)
+                        };
+                        let ds = if s < 0.0 {
+                            xn * s.powi(4) / (1.0 + s)
+                        } else {
+                            xn * s.powi(4)
+                        };
+                        let dt = if t < 0.0 {
+                            xk * t.powi(4) / (1.0 + t)
+                        } else {
+                            xk * t.powi(4)
+                        };
+                        let de = if e < 0.0 {
+                            nm * e.powi(4) / (1.0 + e)
+                        } else {
+                            nm * e.powi(4)
+                        };
+        
+                        if av < ub - 0.25*(dr + ds + dt + de) + (y + m)*(gl - gu) - 0.0078 {
+                            break y as i64;
+                        }
+        
+                        // Step 4.3: Final Acceptance/Rejection Test
+                        let av_critical = a -
+                            ln_of_factorial(y) -
+                            ln_of_factorial(n1 as f64 - y) - 
+                            ln_of_factorial(k as f64 - y) - 
+                            ln_of_factorial((n2 - k) as f64 + y);
+                        if v.ln() <= av_critical {
+                            break y as i64;
+                        }
+                    }
+                }
+            }
+        };
+
+        (offset_x + sign_x * x) as u64
```

Comment:
I like your notation more than the one in the reference! I don't have time to verify it right now, but will do so later.

Reply:
I got this wrong initially, so I made sure the tests cover all four cases from the paper :)

Reply:
Ok, all four cases check out!

Sorry it took me a while.